pkg/cmd/infra/pusher/pusher.go

```diff
@@ -0,0 +1,392 @@
+package pusher
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"strings"
+
+	"github.com/docker/distribution"
+	"github.com/docker/distribution/digest"
+	"github.com/docker/distribution/reference"
+	"github.com/docker/distribution/registry/client"
+	"github.com/docker/distribution/registry/client/auth"
+	"github.com/golang/glog"
+	"github.com/spf13/cobra"
+
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+	apirequest "k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/kubernetes/pkg/kubectl/cmd/templates"
+	kcmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+
+	"github.com/docker/distribution/manifest/schema2"
+	ocmd "github.com/openshift/origin/pkg/cmd/cli/cmd"
+	imageapi "github.com/openshift/origin/pkg/image/api"
+	"github.com/openshift/origin/pkg/image/importer"
+)
+
+var (
+	longDesc = templates.LongDesc(`
+		Push an image to a new location
```

Comment:
This is still draft, but the use case is going to be a controller that runs a pod in your namespace that syncs your images to other repositories.  So this is just the first part, still need to add some other bits:

1. make it easy to inject pull secrets that can be used
2. the controller itself
3. figure out how to batch this - I kind of think we want to batch all of the "syncs" in a namespace all at once into one job.